src/v/config/property.h

```diff
@@ -593,4 +596,33 @@ class enum_property : public property<T> {
     std::vector<T> _values;
 };
 
+class retention_duration_property final
+  : public property<std::optional<std::chrono::milliseconds>> {
+public:
+    using property<std::optional<std::chrono::milliseconds>>::property;
+
+    void set_value(std::any v) override {
+        auto value = std::any_cast<std::chrono::milliseconds>(std::move(v));
+        property<std::optional<std::chrono::milliseconds>>::update_value(
+          std::optional(value));
+    }
+
+    bool set_value(YAML::Node n) override {
+        auto value = n.as<std::chrono::milliseconds>();
+        return property<std::optional<std::chrono::milliseconds>>::update_value(
+          std::optional(value));
+    }
+
+    const std::optional<std::chrono::milliseconds> operator()() const {
```

Comment:
This is a bit tricky. This behaviour is only customised if the concrete type is known statically, not if it's used polymorphically. This customisation also doesn't apply to:
* `const T& value()`
* `const T& value() const`
* `const T& operator()()`
* `operator T() const`

I don't think customising the "read" side of property is possibly currently. I think it has to be done on the "write" side, via `set_value`.

Reply:
got it, will fix